From a8166eaba6d66fcb030de04c81c1d78fd304c860 Mon Sep 17 00:00:00 2001
From: hyphop <art@khadas.com>
Date: Sat, 16 May 2020 17:04:52 +0900
Subject: [PATCH] hdmi_force_activate_and_failsave_4k_to_fullhd

+ no wait_for_hpd and force hdmi init
+ 4K failsave to fullhd

---
 drivers/video/rockchip/rk_hdmi.c |  5 +++--
 drivers/video/rockchip/rk_vop.c  | 17 ++++++++++++++---
 2 files changed, 17 insertions(+), 5 deletions(-)

diff --git a/drivers/video/rockchip/rk_hdmi.c b/drivers/video/rockchip/rk_hdmi.c
index 5b44a7e8..4c543c1b 100644
--- a/drivers/video/rockchip/rk_hdmi.c
+++ b/drivers/video/rockchip/rk_hdmi.c
@@ -113,8 +113,9 @@ int rk_hdmi_probe(struct udevice *dev)
 
 	ret = dw_hdmi_phy_wait_for_hpd(hdmi);
 	if (ret < 0) {
-		debug("hdmi can not get hpd signal\n");
-		return -1;
+		debug("*** hdmi can not get hpd signal *** \n");
+		debug("*** force hdmi activate  *** \n");
+	//	return -1;
 	}
 
 	dw_hdmi_init(hdmi);
diff --git a/drivers/video/rockchip/rk_vop.c b/drivers/video/rockchip/rk_vop.c
index e1bd656b..44765bbc 100644
--- a/drivers/video/rockchip/rk_vop.c
+++ b/drivers/video/rockchip/rk_vop.c
@@ -299,9 +299,20 @@ static int rk_display_init(struct udevice *dev, ulong fbbase, ofnode ep_node)
 	}
 
 	ret = display_read_timing(disp, &timing);
-	if (ret) {
-		debug("%s: Failed to read timings\n", __func__);
-		return ret;
+
+	if (ret || timing.hactive.typ > 1920 ) {
+	    debug("%s: try failsave - full hd mode timing\n", __func__);
+	    timing.pixelclock.typ = 148500000;
+	    timing.hactive.typ = 1920;
+	    timing.hfront_porch.typ = 88;
+	    timing.hback_porch.typ = 148;
+	    timing.hsync_len.typ = 44;
+	    timing.vactive.typ = 1080;
+	    timing.vfront_porch.typ = 4;
+	    timing.vback_porch.typ = 36;
+	    timing.vsync_len.typ = 5;
+	    timing.flags = 10;
+	    timing.hdmi_monitor = true;
 	}
 
 	ret = clk_get_by_index(dev, 1, &clk);
-- 
2.17.1

